### ISSUE TITLE
Issue with highlighting and using multi-words synonyms

### ISSUE THREAD

**Issue by aslamy** [elastic/elasticsearch#26497]
Like it said here https://github.com/elastic/elasticsearch/issues/3982 highlighting working weirdly with synonym filter. The problem occurs when multi-word synonyms are in use. 
I thought it's worth to make a new request for this issue.

**Comment by aslamy** [comment]
I have synonym token filter like : "united kingdom , uk => united kingdom , uk".
When I use my analyser then I get wierd tokens:
```
{
  "analyzer" : "custom_analyzer",
  "text" : ["UK comprises four countries"]
}
{
	"tokens": [
		{
			"token": "united",
			"start_offset": 0,
			"end_offset": 2,
			"type": "SYNONYM",
			"position": 0
		},
		{
			"token": "uk",
			"start_offset": 0,
			"end_offset": 2,
			"type": "SYNONYM",
			"position": 0
		},
		{
			"token": "comprises",
			"start_offset": 3,
			"end_offset": 12,
			"type": "<ALPHANUM>",
			"position": 1
		},
		{
			"token": "kingdom",
			"start_offset": 3,
			"end_offset": 12,
			"type": "SYNONYM",
			"position": 1
		},
		{
			"token": "four",
			"start_offset": 13,
			"end_offset": 17,
			"type": "<ALPHANUM>",
			"position": 2
		},
		{
			"token": "countries",
			"start_offset": 18,
			"end_offset": 27,
			"type": "<ALPHANUM>",
			"position": 3
		}
	]
}
```
Token comprises and kingdom has incorrect order.  Could this be the reason that the highlight does not work properly?

**Comment by jimczi** [comment]
> Could this be the reason that the highlighter does not work properly?

Yes, the default `synonym` filter is not able to handle multi-word synonyms expansion correctly. 
In 5.x a new filter called `synonym_graph` has been added, it works correctly with multi-words synonym expansion:
https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-graph-tokenfilter.html
It must be applied at query time only (`search_analyzer`) and fixes issues when highlighting multi-word synonyms.
I'll add a note in the documentation of the `synonym` filter.

**Comment by aslamy** [comment]
Thank you !

**Comment by aslamy** [comment]
I tested synonym_graph. Highlighting multi-word synonyms works but now results are returned in different relevance for synonyms. Should not synonyms give exactly same documents as result?

**Comment by weberhofer** [comment]
This issue might be related to and been fixed with #26097 in version 6.1.

**Comment by romseygeek** [comment]
Fixed by #26097
